libraries/chain/apply_context.cpp

```diff
@@ -626,10 +627,11 @@ void apply_context::schedule_deferred_transaction( const uint128_t& sender_id, a
 }
 
 bool apply_context::cancel_deferred_transaction( const uint128_t& sender_id, account_name sender ) {
+   EOS_ASSERT( !trx_context.is_read_only(), transaction_exception, "cannot cancel a readonly deferred transaction" );
```

Comment:
A better error message would be `"cannot cancel a deferred transaction from within a readonly transaction"`.